### ISSUE TITLE
write nn.Module with attention inside

### ISSUE THREAD

**Issue by sushruta** [sushruta/pytorch-examples#4]
write `forward`, `backward`, `step`, etc. functions from scratch with vanilla attention

Ideally, the result should be something like AttentionModule where the forward, backward, step, optimizer, etc. functions are all written by hand.

To test this, we will write a module with only forward function. We will integrate these two modules in a test module and check if the weights after every step are essentially the same (subject to a small epsilon). We can also look at validation loss and make sure it is the same/converging to declare that the implementation is correct.
